src/metabase/api/dashboard.clj

```diff
@@ -580,37 +646,58 @@
                                                                [:target       :any]]]]]
    [:series             {:optional true} [:maybe [:sequential map?]]]])
 
+(def ^:private UpdatedDashboardTab
+  [:map
+   ;; id can be negative, it indicates a new card and BE should create them
+   [:id   ms/Int]
+   [:name ms/NonBlankString]])
+
 (api/defendpoint PUT "/:id/cards"
-  "Update `Cards` on a Dashboard. Request body should have the form:
-
-    {:cards [{:id                 ... ; DashboardCard ID
-              :size_x             ...
-              :size_y             ...
-              :row                ...
-              :col                ...
-              :parameter_mappings ...
-              :series             [{:id 123
-                                    ...}]}
-             ...]}"
-  [id :as {{:keys [cards]} :body}]
-  {id    ms/PositiveInt
-   cards [:sequential UpdatedDashboardCard]}
-  (let [dashboard     (-> (api/write-check Dashboard id)
-                          api/check-not-archived
-                          (t2/hydrate [:ordered_cards :series :card]))
-        current-cards (:ordered_cards dashboard)
-
-        {:keys [to-update to-delete to-create]} (classify-changes current-cards cards)]
+  "Update `Cards` and `Tabs` on a Dashboard. Request body should have the form:
+
+    {:cards        [{:id                 ... ; DashboardCard ID
+                     :size_x             ...
+                     :size_y             ...
+                     :row                ...
+                     :col                ...
+                     :parameter_mappings ...
+                     :series             [{:id 123
+                                           ...}]}
+                     ...]
+     :ordered_tabs [{:id       ... ; DashboardTab ID
+                     :name     ...}]}"
+  [id :as {{:keys [cards ordered_tabs]} :body}]
+  {id           ms/PositiveInt
+   cards        (ms/maps-with-unique-key [:sequential UpdatedDashboardCard] :id)
+   ordered_tabs (ms/maps-with-unique-key [:sequential UpdatedDashboardTab] :id)}
+  (let [dashboard (-> (api/write-check Dashboard id)
+                      api/check-not-archived
+                      (t2/hydrate [:ordered_cards :series :card] :ordered_tabs))
+        new-tabs  (when-not (nil? ordered_tabs)
+                    (map-indexed (fn [idx tab] (assoc tab :position idx)) ordered_tabs))]
```

Comment:
ordered_tabs is non-nil now.
```suggestion
        new-tabs  (map-indexed (fn [idx tab] (assoc tab :position idx)) ordered_tabs)]
```

Reply:
removed in [df19bd2](https://github.com/metabase/metabase/pull/29861/commits/df19bd2bd89325e5cbfd3531ec177296bbd8f8b5)